EIPS/eip-7979.md

```diff
@@ -0,0 +1,436 @@
+
+Worse, jumps that can dynamically branch to any destination in the program can be a denial-of-service vulnerability — they can cause quadratic "path explosions" when traversing the program's flow of control.
+
+#### Tracing control flow
+
+A fundamental operation on computer programs is to trace the complete flow of control of code sequences in a minimal number of steps.  This can be done by starting at the first instruction, tracing instruction by instruction until a jump is encountered, and recursively tracing the possible destinations of each jump; when a previously encountered jump destination is reached the recursion returns.  When all jumps are static the number of steps is linear in the number of instructions — only one or two paths must be explored at each jump.  With dynamic jumps every possible path must be explored at each jump, leading to possible "path explosions" — the number of steps can be quadratic in the number of instructions.
+
+#### Denial of service
+
+For Ethereum, quadratic tracing times are an online denial-of-service vulnerability that prevents us — in time and space linear in the size of the code — from validating the safe use of EVM code, or transforming it for faster execution, whether by an interpreter, JIT, or AOT compiler.
+
+Even offline, many static analyses can become impractically slow in depth and at scale.  This include verifying code against formal specifications and other proofs of program invariants —  with so much at stake on the blockchain correct contracts are crucial.  These analyses must first get past the difficulties of tracing control flow (_Analyzing Smart Contracts: From EVM to a sound Control-Flow Graph._) and one proposal goes so far as using neural networks to recover structure from code, _Neural-FEBI: Accurate Function Identification in Ethereum Virtual Machine Bytecode._  This proposal renders these problems trivial.
+
+#### Taming the EVM
+
+To prevent tracing from "going quadratic" we must prevent the dynamic use of jumps.  Most JUMP and JUMPI instructions in EVM code are preceded by a `PUSH` — that is, they are static.  So this proposal can validate that `JUMP` and `JUMPI` are used statically.   The only places that jumps _must_ be used dynamically are to support calls to Turing's _subsidiary operations._  For that purpose we propose `JUMPSUB` and `RETURNSUB` opcodes as replacements. 
+
+## Specification
+
+### `JUMPSUB (0x5f)`
+
+Transfers control to a subsidiary operation.
+
+1. Pop the `destination` from the `data stack`.
+2. Push the current `PC + 1` to the `return stack`.
+3. Set `PC` to `destination`.
```

Comment:
In non-valid code what if we `JUMPSUB` to a non-`JUMPDEST`? Will this exceptionally abort or not? :thinking: 

Should this validate the return stack size? It seems in valid code there is a max of 1024 items.